A couple of toolchain TLS fixes

keith-packard opened the pull request:
The first recognizes that crosstool-ng enables TLS by default, and so Zephyr should at least allow TLS to be enabled when building with it.

The second patch allows any toolchain used with the 'cross-compile' variant to be configured to support thread local storage (disabled by default).

With these two patches, I am able to build Zephyr projects using the 'arm-none-eabi' sample configuration provided with crosstool-ng (save for GCC 12 issues, of course), the Debian arm-none-eabi toolchain and the regular Zephyr arm-zephyr-eabi crosstools-ng toolchain.

keith-packard committed:
toolchain/xtools: Crosstools-ng supports TLS by default

The Zephyr cross tools configuration which the 'xtools' variant supports
enables TLS, so set TOOLCHAIN_SUPPORTS_THREAD_LOCAL_STORAGE when using it.

Signed-off-by: Keith Packard <keithp@keithp.com>

tejlmand commented on cmake/toolchain/xtools/target.cmake:
```diff
@@ -13,6 +13,8 @@ set(CROSS_COMPILE_TARGET_sparc     sparc-zephyr-elf)
 set(CROSS_COMPILE_TARGET ${CROSS_COMPILE_TARGET_${ARCH}})
 set(SYSROOT_TARGET       ${CROSS_COMPILE_TARGET})
 
+set(ZEPHYR_TOOLCHAIN_SUPPORTS_THREAD_LOCAL_STORAGE y)
```
this doesn't set the variable in environment.

also this commit brought my attention to this PR: #45656

We should not in Kconfig rely directly on environment settings define by the user, as it can result in incremental builds starting to behave differently.

Instead such knowledge should come from the toolchains Kconfig like here: 
https://github.com/zephyrproject-rtos/zephyr/blob/1dc69c064821d3d37f80b10c7ed15645ea543fce/cmake/toolchain/gnuarmemb/Kconfig#L8

environment variables to be used inside Kconfig should be exported by CMake kconfig integration here:
https://github.com/zephyrproject-rtos/zephyr/blob/1dc69c064821d3d37f80b10c7ed15645ea543fce/cmake/modules/kconfig.cmake#L110

but for the TLS, I believe the first approachh would be best.
Done correctly, it could also provide users an option to state that a toolchain doesn't support TLS.

tejlmand commented on cmake/toolchain/xtools/target.cmake:
```diff
@@ -10,7 +10,11 @@ set(CROSS_COMPILE_TARGET_arc         arc-zephyr-elf)
 set(CROSS_COMPILE_TARGET_sparc     sparc-zephyr-elf)
 
-set(CROSS_COMPILE_TARGET ${CROSS_COMPILE_TARGET_${ARCH}})
+if(DEFINED ENV{ZEPHYR_CROSS_COMPILE_TARGET})
+  set(CROSS_COMPILE_TARGET "$ENV{ZEPHYR_CROSS_COMPILE_TARGET}")
+else()
+  set(CROSS_COMPILE_TARGET ${CROSS_COMPILE_TARGET_${ARCH}})
+endif()
```
This seems to be adding to this mess: https://github.com/zephyrproject-rtos/zephyr/issues/40389

Also there is no documentation of this variable, so how are users expected to know about it ?

keith-packard replied:
Naming suggestions welcome; as you say, there isn't much consistency here. I figured that using a ZEPHYR_ prefix was a bit more in line with similar usage elsewhere. However, reviewing the crosstool-ng docs, it seems like using XTOOLS_ would be a better choice

I've changed the name to match the other crosstools-ng environment variable names and also included a patch for the docs.

tejlmand commented on cmake/toolchain/xtools/target.cmake:
```diff
@@ -10,7 +10,11 @@ set(CROSS_COMPILE_TARGET_arc         arc-zephyr-elf)
 set(CROSS_COMPILE_TARGET_sparc     sparc-zephyr-elf)
 
-set(CROSS_COMPILE_TARGET ${CROSS_COMPILE_TARGET_${ARCH}})
+if(DEFINED ENV{XTOOLS_CROSS_COMPILE_TARGET})
+  set(CROSS_COMPILE_TARGET "$ENV{XTOOLS_CROSS_COMPILE_TARGET}")
+else()
```
your commit message says:
> When crosstool-ng is used as a build system, the target triple might not
include 'zephyr' -- arm-none-eabi is quite commonly used for compatibility
with the GNU Arm Embedded Toolkit.

but if you read the guide here: https://docs.zephyrproject.org/latest/develop/toolchains/crosstool_ng.html
> Follow the [Zephyr SDK with Crosstool NG instructions](https://github.com/zephyrproject-rtos/sdk-ng/blob/master/README.md) to build your toolchain. Repeat as necessary to build toolchains for multiple target architectures.

users are expected to use the Zephyr SDK build instructions for `xtools`.
If users are manually building using cross-tool they could even change more than just the triplet.
They could even set `CT_CC_GCC_CONFIG_TLS=n`.

Because of this, and because we already support a generic `CROSS_COMPILE` setting:
https://docs.zephyrproject.org/latest/develop/toolchains/other_x_compilers.html
Users with a different triplet can just do:
```
$ export ZEPHYR_TOOLCHAIN_VARIANT=cross-compile
$ export CROSS_COMPILE=/my/gcc/arm-none-eabi-
```

so I fail to see why we need to have more customization in `xtools` when they users can already accomplish exactly this, using `cross-compile`.
Let's not introduce more ways of achieving the same.

Especially since the `ZEPHYR_TOOLCHAIN_VARIANT=xtools` is a convenience that selects the exact toolchain based on the arch of the target, whereas the `XTOOLS_CROSS_COMPILE_TARGET` needs to be adjusted by the user when switching from `arm` to `xtensa`.
Thus the proposed solution in this commit is directly comparable to the use of `CROSS_COMPILE` where the arch is part of the triplet set by the user.

keith-packard replied:
Thanks for your explanation; I hadn't understood the difference between the xtools and cross-compile variants.

tejlmand replied:
And thanks for taking time to clean up this :+1:

tejlmand commented on doc/develop/toolchains/crosstool_ng.rst:
```diff
@@ -25,6 +25,8 @@ You can build toolchains from source code using crosstool-NG.
 
    - Set :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` to ``xtools``.
    - Set :envvar:`XTOOLS_TOOLCHAIN_PATH` to the toolchain build directory.
+   - Set :envvar:`XTOOLS_CROSS_COMPILE_TARGET` to the target triplet of the toolchain,
+     e.g. ``arm-zephyr-eabi``.
```
maybe a better option is to add a note, that if you're customizing the crosstool built toolchain (compared to Zephyr setup) then follow the cross compile toolchain setup.

keith-packard replied:
Yup, and that all works now. The second patch in this series allows a cross-compile variant toolchain to select TLS if it is supported. Thanks much for your help.

tejlmand reviewed (changes requested):
I think the first commit is valuable but should be cleaned up, so that the Kconfig goes into the toolchain folder.

Second commit is not really needed, as this functionality is already available to users if following:
https://docs.zephyrproject.org/latest/develop/toolchains/other_x_compilers.html

keith-packard committed:
toolchain/cross-compile: Allow TLS support selection.

Add a user-settable configuration value indicating whether the toolchain
selected by the cross-compile variant supports thread local storage, using
the default 'n' value to preserve compatibility with existing uses.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented on kernel/Kconfig:
```diff
@@ -922,7 +922,7 @@ config TICKLESS_KERNEL
 config TOOLCHAIN_SUPPORTS_THREAD_LOCAL_STORAGE
 	bool
-	default y if "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr" || "$(ZEPHYR_TOOLCHAIN_SUPPORTS_THREAD_LOCAL_STORAGE)" = "y"
+	default y if "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr" || "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "xtools" || "$(ZEPHYR_TOOLCHAIN_SUPPORTS_THREAD_LOCAL_STORAGE)" = "y"
```
oh! Thanks much for the pointers. I've changed this series to only adjust TLS support for both the xtools and cross-compile toolchains. I've also posted a PR to the Zephyr toolchain to add an explicit indication of TLS support  here: https://github.com/zephyrproject-rtos/sdk-ng/pull/557 If that is merged, we could eventually remove the explicit check for the zephyr variant in kernel/Kconfig, but there's no hurry.

tejlmand replied:
thanks for adding the possibility for cross-compile users to specify whether it supports TLS.

tejlmand commented on cmake/toolchain/cross-compile/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config TOOLCHAIN_CROSS_COMPILE_SUPPORTS_THREAD_LOCAL_STORAGE
+	prompt "Cross-compile toolchain supports TLS"
+	def_bool n
```
default value for a bool is `n`, so `bool` is enough, which again allows you to skip the `prompt` part.

Can be simplified to:
```suggestion
	bool "Cross-compile toolchain supports TLS"
```

Ref: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#redundant-defaults

keith-packard replied:
thanks, that's a bit cleaner looking.

tejlmand reviewed (commented):
looks good now.

A minor nit that would be nice to get fixed.

stephanosio commented on cmake/toolchain/xtools/Kconfig:
```diff
@@ -0,0 +1,6 @@
+# Copyright © 2022 Keith Packard
+# SPDX-License-Identifier: Apache-2.0
+
+config TOOLCHAIN_XTOOLS_SUPPORTS_THREAD_LOCAL_STORAGE
+	def_bool y
+	select TOOLCHAIN_SUPPORTS_THREAD_LOCAL_STORAGE
```
Is this intentionally promptless? (shouldn't users be able to override this?)

If not:

```kconfig
config TOOLCHAIN_XTOOLS_SUPPORTS_THREAD_LOCAL_STORAGE
	bool "Crosstool-ng toolchain supports TLS"
	default y
	select TOOLCHAIN_SUPPORTS_THREAD_LOCAL_STORAGE
```

tejlmand replied:
no, cause if you follow the guide you'll be building cross-tools according to how we build Zephyr SDK:
https://docs.zephyrproject.org/latest/develop/toolchains/crosstool_ng.html

and then TLS will be supported.

btw. @stephanosio the guide still refers to use `./go.sh <arch>` but that has been removed in the sdk-ng repo, so what is the right procedure now ?

stephanosio replied:
Looks like the whole guide needs to be reworked. crosstool-ng is a standalone (independent from both Zephyr and Zephyr SDK) component and the toolchain build configuration can be whatever user chooses.

I am not sure if there is any value in keeping the `xtools` toolchain variant because you may as well be using the `cross-compile`.

stephanosio replied:
I will post a separate PR removing the `xtools` toolchain variant since it does nothing.

keith-packard replied:
That seems quite reasonable -- I've been testing these patches with crosstool-ng toolchains (both arm-none-eabi and arm-zephyr-eabi) as well as the native Debian arm-none-eabi toolchain.

tejlmand replied:
> I am not sure if there is any value in keeping the xtools toolchain variant because you may as well be using the cross-compile.

@stephanosio and @keith-packard the biggest value I see is that if using the Zephyr crosstool setup, you'll get the `zephyr` triplet and the arch selection of compiler.
With `cross-compile` you must adjust `CROSS_COMPILE` variable when building for a new architecture.

The `xtools` CMake files also serves as a nice template for those creating custom toolchains for many archs, as they may easily see an example of how automatic selection of arch a specific toolchain can be done.

So `xtools` is a conveniance and template, but everything can be achieved using `cross-compile`, just as `ZEPHYR_TOOLCHAIN_VARIANT=zephyr` is a convenience.

We could request users to point to their Zephyr SDK using `export CROSS_COMPILE=/opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-`, but we don't, cause we want an easy way to use Zephyr SDK.

stephanosio replied:
"Zephyr crosstool" is not really a thing. If you are using Zephyr toolchains from the Zephyr SDK, then you would be using the `zephyr` toolchain variant. If not, you will be using the `cross-compile` toolchain variant.

crosstool-ng can build any toolchain from `aarch64-none-elf` to `aarch64-linux-gnu` -- the `xtools` toolchain variant, as it is implemented and stated in the docs, does not really make much sense.

tejlmand reviewed (approved):
lgtm, thanks for the improvement and the patience :+1:
